File: examples/chapter20/e03_microbatching/python/main.py
"""Example 20.3 - Micro-batching (Python reference + parity test).

32 client threads x 40 requests against two servers: one forward pass per
request, and a queue + worker that runs requests in batches of up to 32.

Run:   python main.py
Test:  pytest main.py
"""

import importlib.util
import json
import os
import time

import torch

HERE = os.path.dirname(os.path.abspath(__file__))
WEIGHTS = os.path.join(HERE, "..", "..", "..", "chapter19", "weights")

# Chapter 19's PyTorch model and data generator.
_spec = importlib.util.spec_from_file_location("make_weights", os.path.join(WEIGHTS, "make_weights.py"))
mw = importlib.util.module_from_spec(_spec)
_spec.loader.exec_module(mw)


def load_model():
    from safetensors.torch import load_file
    model = mw.TorchCnn()
    model.load_state_dict(load_file(os.path.join(WEIGHTS, "small_cnn.safetensors")))
    return model.eval()


def images(n):
    rng = mw.Lcg(7)
    return torch.stack([torch.from_numpy(mw.make_image(i % 4, rng)).reshape(1, 12, 12)
                        for i in range(n)])


import queue
import threading
import urllib.request
from concurrent.futures import ThreadPoolExecutor
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

MODEL = load_model()
JOBS = queue.Queue()
BATCHES = [0]


def batch_worker(max_batch=32, max_wait=0.002):
    while True:
        jobs = [JOBS.get()]
        deadline = time.perf_counter() + max_wait
        while len(jobs) < max_batch and time.perf_counter() < deadline:
            try:
                jobs.append(JOBS.get_nowait())
            except queue.Empty:
                time.sleep(0)
        x = torch.stack([torch.tensor(p).reshape(1, 12, 12) for p, _ in jobs])
        with torch.no_grad():
            classes = MODEL(x).argmax(1).tolist()
        for (_, box), c in zip(jobs, classes):
            box["class"] = c
            box["done"].set()
        BATCHES[0] += 1


def make_handler(batched):
    class Handler(BaseHTTPRequestHandler):
        def log_message(self, *args):
            pass

        def do_POST(self):
            pixels = json.loads(self.rfile.read(int(self.headers["Content-Length"])))["pixels"]
            if batched:
                box = {"done": threading.Event()}
                JOBS.put((pixels, box))
                box["done"].wait()
                c = box["class"]
            else:
                with torch.no_grad():
                    c = MODEL(torch.tensor(pixels).reshape(1, 1, 12, 12)).argmax(1).item()
            body = json.dumps({"class": c}).encode()
            self.send_response(200)
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)
    return Handler


class Server(ThreadingHTTPServer):
    request_queue_size = 128   # the default backlog of 5 drops connections under load


def load_test(batched, clients=32, per_client=40):
    server = Server(("127.0.0.1", 0), make_handler(batched))
    threading.Thread(target=server.serve_forever, daemon=True).start()
    url = f"http://127.0.0.1:{server.server_port}/predict"

    def client(c):
        rng, out = mw.Lcg(7 + c), []
        for i in range(per_client):
            body = json.dumps({"pixels": mw.make_image(i % 4, rng).tolist()}).encode()
            t = time.perf_counter()
            with urllib.request.urlopen(urllib.request.Request(url, body)) as r:
                cls = json.loads(r.read())["class"]
            out.append(((time.perf_counter() - t) * 1e3, cls == i % 4))
        return out

    start = time.perf_counter()
    with ThreadPoolExecutor(clients) as pool:
        results = [r for rs in pool.map(client, range(clients)) for r in rs]
    seconds = time.perf_counter() - start
    server.shutdown()
    lat = sorted(ms for ms, _ in results)
    return {"rps": len(lat) / seconds, "p50": lat[int((len(lat) - 1) * 0.5)],
            "p99": lat[int((len(lat) - 1) * 0.99)], "correct": sum(ok for _, ok in results),
            "requests": len(lat)}


def build():
    threading.Thread(target=batch_worker, daemon=True).start()
    a = load_test(batched=False)
    BATCHES[0] = 0
    b = load_test(batched=True)
    return a, b, b["requests"] / max(BATCHES[0], 1)


def main():
    a, b, mean_batch = build()
    print("32 clients x 40 requests each")
    print("strategy   requests/s  p50 ms  p99 ms  correct")
    for name, r in (("direct", a), ("batched", b)):
        print(f"{name:<9}  {r['rps']:>10.0f}  {r['p50']:>6.2f}  {r['p99']:>6.2f}  "
              f"{r['correct']}/{r['requests']}")
    print(f"mean batch size in the batched run = {mean_batch:.1f}")


def test_both_strategies_answer_correctly():
    a, b, _ = build()
    assert (a["correct"], b["correct"]) == (1280, 1280)


if __name__ == "__main__":
    main()

File: examples/chapter20/e03_microbatching/rust/src/main.rs
use std::sync::Arc;
use std::time::{Duration, Instant};

use axum::Router;
use axum::extract::{Json, State};
use axum::routing::post;
use burn::backend::NdArray;
use burn::module::Module;
use burn::nn::conv::{Conv2d, Conv2dConfig};
use burn::nn::pool::{AdaptiveAvgPool2d, AdaptiveAvgPool2dConfig, MaxPool2d, MaxPool2dConfig};
use burn::nn::{Linear, LinearConfig, PaddingConfig2d};
use burn::tensor::activation::relu;
use burn::tensor::backend::Backend;
use burn::tensor::{Tensor, TensorData};
use burn_store::{ModuleSnapshot, PyTorchToBurnAdapter, SafetensorsStore};
use serde::{Deserialize, Serialize};
use tokio::sync::{mpsc, oneshot};

const SIZE: usize = 12;

// --- the data: 12x12 "shapes", generated in code (identical in python.py) ---
// A tiny LCG so Rust and Python draw exactly the same "random" numbers.
struct Lcg(u32);
impl Lcg {
    fn next(&mut self) -> u32 {
        self.0 = self.0.wrapping_mul(1103515245).wrapping_add(12345);
        (self.0 >> 16) & 0x7FFF // 15 bits
    }
    fn below(&mut self, n: u32) -> usize {
        (self.next() % n) as usize
    }
}

// class 0 = horizontal bar, 1 = vertical bar, 2 = diagonal, 3 = plus sign
fn make_image(class: usize, rng: &mut Lcg) -> Vec<f32> {
    let mut img = vec![0.0f32; SIZE * SIZE];
    match class {
        0 => {
            let (r, c) = (rng.below(12), rng.below(7));
            (0..6).for_each(|j| img[r * SIZE + c + j] = 1.0);
        }
        1 => {
            let (c, r) = (rng.below(12), rng.below(7));
            (0..6).for_each(|i| img[(r + i) * SIZE + c] = 1.0);
        }
        2 => {
            let (r, c) = (rng.below(7), rng.below(7));
            (0..6).for_each(|i| img[(r + i) * SIZE + c + i] = 1.0);
        }
        _ => {
            let (r, c) = (2 + rng.below(8), 2 + rng.below(8));
            for d in 0..5 {
                img[r * SIZE + c + d - 2] = 1.0;
                img[(r + d - 2) * SIZE + c] = 1.0;
            }
        }
    }
    // background noise in [0, 0.25): exact in f32, so both languages agree bit for bit
    img.iter_mut()
        .for_each(|p| *p += rng.next() as f32 / 32768.0 * 0.25);
    img
}

// --- the model: Chapter 15's CNN, weights from Chapter 19's safetensors file ---
#[derive(Module, Debug)]
struct SmallCnn<B: Backend> {
    conv1: Conv2d<B>,
    pool: MaxPool2d,
    conv2: Conv2d<B>,
    gap: AdaptiveAvgPool2d,
    head: Linear<B>,
}

impl<B: Backend> SmallCnn<B> {
    fn new(device: &B::Device) -> Self {
        let same = PaddingConfig2d::Explicit(1, 1, 1, 1);
        Self {
            conv1: Conv2dConfig::new([1, 4], [3, 3])
                .with_padding(same.clone())
                .init(device),
            pool: MaxPool2dConfig::new([2, 2]).init(),
            conv2: Conv2dConfig::new([4, 8], [3, 3])
                .with_padding(same)
                .init(device),
            gap: AdaptiveAvgPool2dConfig::new([1, 1]).init(),
            head: LinearConfig::new(8, 4).init(device),
        }
    }

    fn forward(&self, x: Tensor<B, 4>) -> Tensor<B, 2> {
        let x = self.pool.forward(relu(self.conv1.forward(x)));
        let x = relu(self.conv2.forward(x));
        let [batch, channels, _, _] = x.dims();
        self.head
            .forward(self.gap.forward(x).reshape([batch, channels]))
    }
}

const WEIGHTS: &str = concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/../../../chapter19/weights/small_cnn.safetensors"
);

fn load_model<B: Backend>(device: &B::Device) -> SmallCnn<B> {
    let mut store = SafetensorsStore::from_file(WEIGHTS)
        .with_from_adapter(PyTorchToBurnAdapter)
        .with_key_remapping(r"^features\.0\.(.+)$", "conv1.$1")
        .with_key_remapping(r"^features\.3\.(.+)$", "conv2.$1")
        .with_key_remapping(r"^classifier\.(.+)$", "head.$1");
    let mut model = SmallCnn::new(device);
    model.load_from(&mut store).expect("load weights");
    model
}

#[derive(Deserialize, Serialize)]
struct PredictRequest {
    pixels: Vec<f32>,
}

#[derive(Deserialize, Serialize)]
struct PredictResponse {
    class: usize,
}

fn argmax(row: &[f32]) -> usize {
    (0..row.len())
        .max_by(|&a, &b| row[a].total_cmp(&row[b]))
        .unwrap()
}

// ---- strategy A: every request runs its own forward pass (on a blocking thread) ----
async fn predict_direct(
    State(model): State<Arc<SmallCnn<NdArray>>>,
    Json(req): Json<PredictRequest>,
) -> Json<PredictResponse> {
    let class = tokio::task::spawn_blocking(move || {
        let x = Tensor::from_data(
            TensorData::new(req.pixels, [1, 1, SIZE, SIZE]),
            &Default::default(),
        );
        argmax(&model.forward(x).into_data().to_vec::<f32>().unwrap())
    })
    .await
    .unwrap();
    Json(PredictResponse { class })
}

// ---- strategy B: requests queue up; one worker runs them in batches ----
type Job = (Vec<f32>, oneshot::Sender<usize>);

fn spawn_batcher(
    model: SmallCnn<NdArray>,
    max_batch: usize,
    max_wait: Duration,
) -> mpsc::Sender<Job> {
    let (tx, mut rx) = mpsc::channel::<Job>(1024);
    std::thread::spawn(move || {
        let device = Default::default();
        // Block until the first request arrives, then collect more for at most max_wait.
        while let Some(first) = rx.blocking_recv() {
            let mut jobs = vec![first];
            let deadline = Instant::now() + max_wait;
            while jobs.len() < max_batch && Instant::now() < deadline {
                match rx.try_recv() {
                    Ok(job) => jobs.push(job),
                    Err(_) => std::thread::yield_now(),
                }
            }
            let n = jobs.len();
            let pixels: Vec<f32> = jobs.iter().flat_map(|(p, _)| p.iter().copied()).collect();
            let x = Tensor::<NdArray, 4>::from_data(
                TensorData::new(pixels, [n, 1, SIZE, SIZE]),
                &device,
            );
            let logits: Vec<f32> = model.forward(x).into_data().to_vec().unwrap();
            for ((_, reply), row) in jobs.into_iter().zip(logits.chunks(4)) {
                let _ = reply.send(argmax(row));
            }
            BATCHES.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
        }
    });
    tx
}
static BATCHES: std::sync::atomic::AtomicUsize = std::sync::atomic::AtomicUsize::new(0);

async fn predict_batched(
    State(queue): State<mpsc::Sender<Job>>,
    Json(req): Json<PredictRequest>,
) -> Json<PredictResponse> {
    let (tx, rx) = oneshot::channel();
    queue.send((req.pixels, tx)).await.unwrap();
    Json(PredictResponse {
        class: rx.await.unwrap(),
    })
}

// ---- the load test: `clients` concurrent clients, `per_client` requests each ----
struct Report {
    requests: usize,
    seconds: f64,
    p50_ms: f64,
    p99_ms: f64,
    correct: usize,
}

async fn load_test(app: Router, clients: usize, per_client: usize) -> Report {
    let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
    let url = format!("http://{}/predict", listener.local_addr().unwrap());
    tokio::spawn(async move { axum::serve(listener, app).await.unwrap() });

    let start = Instant::now();
    let tasks: Vec<_> = (0..clients)
        .map(|c| {
            let url = url.clone();
            tokio::spawn(async move {
                let client = reqwest::Client::new();
                let mut rng = Lcg(7 + c as u32);
                let mut out = Vec::new();
                for i in 0..per_client {
                    let class = i % 4;
                    let body = PredictRequest {
                        pixels: make_image(class, &mut rng),
                    };
                    let t = Instant::now();
                    let resp: PredictResponse = client
                        .post(&url)
                        .json(&body)
                        .send()
                        .await
                        .unwrap()
                        .json()
                        .await
                        .unwrap();
                    out.push((t.elapsed().as_secs_f64() * 1e3, resp.class == class));
                }
                out
            })
        })
        .collect();
    let mut latencies = Vec::new();
    let mut correct = 0;
    for t in tasks {
        for (ms, ok) in t.await.unwrap() {
            latencies.push(ms);
            correct += ok as usize;
        }
    }
    let seconds = start.elapsed().as_secs_f64();
    latencies.sort_by(f64::total_cmp);
    let pct = |p: f64| latencies[((latencies.len() - 1) as f64 * p) as usize];
    Report {
        requests: latencies.len(),
        seconds,
        p50_ms: pct(0.50),
        p99_ms: pct(0.99),
        correct,
    }
}

/// Returns (direct report, batched report, mean batch size).
async fn build() -> (Report, Report, f64) {
    let device = Default::default();
    let (clients, per_client) = (32, 40);

    let direct = Router::new()
        .route("/predict", post(predict_direct))
        .with_state(Arc::new(load_model::<NdArray>(&device)));
    let a = load_test(direct, clients, per_client).await;

    let queue = spawn_batcher(load_model::<NdArray>(&device), 32, Duration::from_millis(2));
    let batched = Router::new()
        .route("/predict", post(predict_batched))
        .with_state(queue);
    let b = load_test(batched, clients, per_client).await;
    let mean_batch = b.requests as f64 / BATCHES.load(std::sync::atomic::Ordering::Relaxed) as f64;
    (a, b, mean_batch)
}

#[tokio::main]
async fn main() {
    let (a, b, mean_batch) = build().await;
    println!("32 clients x 40 requests each");
    println!("strategy   requests/s  p50 ms  p99 ms  correct");
    for (name, r) in [("direct", &a), ("batched", &b)] {
        println!(
            "{name:<9}  {:>10.0}  {:>6.2}  {:>6.2}  {}/{}",
            r.requests as f64 / r.seconds,
            r.p50_ms,
            r.p99_ms,
            r.correct,
            r.requests
        );
    }
    println!("mean batch size in the batched run = {mean_batch:.1}");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test(flavor = "multi_thread")]
    async fn both_strategies_answer_correctly() {
        let (a, b, mean_batch) = build().await;
        assert_eq!((a.correct, a.requests), (1280, 1280));
        assert_eq!((b.correct, b.requests), (1280, 1280));
        assert!(mean_batch >= 1.0);
    }
}
